apm-agent-core/src/test/java/co/elastic/apm/agent/logging/LoggingConfigurationTest.java

```diff
@@ -0,0 +1,206 @@
+    static void setup() {
+        ElasticApmTracer tracer = MockTracer.createRealTracer();
+        //noinspection unchecked
+        logLevelConfig = (ConfigurationOption<LogLevel>) tracer.getConfigurationRegistry().getConfigurationOptionByKey("log_level");
+        ElasticApmAgent.initInstrumentation(tracer, ByteBuddyAgent.install(), List.of(new LoggerTestInstrumentation()));
+
+        // We need to clean the current contexts for this test to resemble early agent setup
+        originalLoggerContextFactory = LogManager.getFactory();
+        // Setting with a custom context factory that create a new context for each ClassLoader. This is similar to the default
+        // context log4j2 context factory, with one exception - this simple factory doesn't consider ClassLoader hierarchy
+        // as the context, but each ClassLoader is consider a different context.
+        // This better reflects the runtime environment of real agent, where the agent CL is not part of the CL hierarchy
+        // of plugin class loaders (it's contents are available, but it is not part of the inherent CL hierarchy)
+        testLog4jContextFactory = new TestLog4jContextFactory();
+        LogManager.setFactory(testLog4jContextFactory);
+
+        // Not really an agent logger, but representing the agent-level logger
+        agentLogger = LoggerFactory.getLogger(LoggingConfigurationTest.class);
+    }
+
+    @AfterAll
+    static void reset() {
+        ElasticApmAgent.reset();
+    }
+
+    @AfterEach
+    void tearDown() {
+        // restoring the original logger context factory so that other tests are unaffected
+        LogManager.setFactory(originalLoggerContextFactory);
+    }
+
+    @Test
+    void loggingLevelChangeTest() throws IOException {
+        // Assuming default is debug level in tests based on test.elasticapm.properties
+        assertThat(agentLogger.isTraceEnabled()).isFalse();
+        // A logger created by a plugin CL - see LoggerTestInstrumentation
+        Logger pluginLogger = new LoggerTest().getLogger();
+        assertThat(pluginLogger).isNotNull();
+        LoggerContext pluginLoggerContext = testLog4jContextFactory.getContext(pluginLogger);
+        assertThat(pluginLoggerContext).isNotNull();
+        assertThat(pluginLoggerContext.getName()).startsWith(IndyPluginClassLoader.class.getName());
+        assertThat(pluginLogger.isTraceEnabled()).isFalse();
+
+        logLevelConfig.update(LogLevel.TRACE, SpyConfiguration.CONFIG_SOURCE_NAME);
+        assertThat(agentLogger.isTraceEnabled()).isTrue();
+        assertThat(pluginLogger.isTraceEnabled()).isTrue();
+    }
+
+    private static class LoggerTest {
+        @Nullable
+        Logger getLogger() {
+            return null;
+        }
+    }
+
+    private static final class TestLog4jContextFactory extends Log4jContextFactory {
+
+        ContextSelector contextSelector = new TestContextSelector();
+
+        @Nullable
+        LoggerContext getContext(Logger slf4jLogger) {
```

Comment:
why is this argument named `slf4jLogger` ?

Reply:
Because it's an `org.slf4j.Logger`, as opposed to the `org.apache.logging.log4j.core.Logger` used two lines below. It's a utility for the test that only knows slf4j loggers.